chain/lf3/manifest.go

```
@@ -84,9 +102,212 @@ func NewManifestProvider(mctx helpers.MetricsCtx, config *Config, cs *store.Chai
+		networkName: config.BaseNetworkName,
+		CheckPeriod: 15 * time.Minute,
+
+		manifestChanges: make(chan *manifest.Manifest, 1),
+
+		errgrp:     errgrp,
+		runningCtx: ctx,
+		cancel:     cancel,
+	}, nil
+}
+
+func (cmp *ContractManifestProvider) Start(context.Context) error {
+	// no address, nothing to do
+	if len(cmp.address) == 0 {
+		// close the channel so fusing provider knows we have nothing
+		close(cmp.manifestChanges)
+		return nil
+	}
+
+	m, err := cmp.fetchActivationInfo(cmp.runningCtx)
+	if err != nil {
+		log.Warnw("got error while fetching manifest from contract", "error", err)
+	}
+	cmp.manifestChanges <- m
+
+	cmp.errgrp.Go(func() error {
+		t := time.NewTicker(cmp.CheckPeriod)
+		defer t.Stop()
+
+		var knownManifest *manifest.Manifest
+	loop:
+		for cmp.runningCtx.Err() == nil {
+			select {
+			case <-t.C:
+				m, err := cmp.fetchActivationInfo(cmp.runningCtx)
+				if err != nil {
+					log.Warnw("got error while fetching manifest from contract", "error", err)
+					continue loop
+				}
+
+				if knownManifest.Equal(m) {
+					continue loop
+				}
+
+				if m != nil {
+					log.Infow("new manifest from contract", "enabled", true, "bootstrapEpoch", m.BootstrapEpoch)
+				} else {
+					log.Info("new manifest from contract", "enabled", false)
+				}
+				cmp.manifestChanges <- m
+				knownManifest = m
+			case <-cmp.runningCtx.Done():
+			}
+		}
+
+		return nil
+	})
+	return nil
+}
+
+func (cmp *ContractManifestProvider) fetchActivationInfo(ctx context.Context) (*manifest.Manifest, error) {
+	ethReturn, err := cmp.callContract(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("calling contract at %s: %w", cmp.address, err)
+	}
+	if len(ethReturn) == 0 {
+		return nil, nil
+	}
+
+	activationEpoch, compressedManifest, err := cmp.parseContractReturn(ethReturn)
+
+	if activationEpoch == math.MaxUint64 || len(compressedManifest) == 0 {
+		return nil, nil
+	}
+
+	reader := io.LimitReader(flate.NewReader(bytes.NewReader(compressedManifest)), 1<<20)
```

Why secondary limit reader? The `compressedManifest` was read as byte slice via a private func. Would it make sense to check the length earlier and fail faster before we come to decode it if it is too long?